README.md
# NavigationView (1.0.6)

[![pub package](https://img.shields.io/pub/v/navigation_view.svg)](https://pub.dev/packages/navigation_view)
[![pub points](https://img.shields.io/pub/points/navigation_view?color=2E8B57&label=pub%20points)](https://pub.dev/packages/navigation_view/score)



NavigationView is a Flutter package that provides a navigation component with customizable
colors and icons. It allows you to easily create a navigation bar with animated transitions between
different pages.
by most animation flutter

## Demo new

<img src="https://github.com/PuzzleTakX/navigation_view/blob/master/demo/3.gif?raw=true" alt="image_demo" width="300" height="700">

## Installation ☺

To use the NavigationView package, add the following dependency to your `pubspec.yaml`
file:

[![pub package](https://img.shields.io/pub/v/navigation_view.svg)](https://pub.dev/packages/navigation_view)
```yaml
dependencies: 
navigation_view: ^1.0.6
```
Then, run `flutter pub get` to fetch the package.

## Usage

Import the package in your Dart file:

```dart
import 'package:navigation_view/navigation_view.dart';
```

Create a `NavigationView` widget and provide the necessary parameters:

```dart
 NavigationView(
onChangePage: (c) {},
curve: Curves.easeInBack,
durationAnimation: const Duration(milliseconds: 400),
items: [
    ItemNavigationView(childAfter: const Icon(Icons.home_rounded,color: Colors.blue,size: 30,),
    childBefore: Icon(Icons.home_outlined,color: Colors.grey.withAlpha(60),size: 30,)),
    ItemNavigationView(childAfter: const Icon(Icons.widgets_rounded,color: Colors.blue,size: 30,),
    childBefore: Icon(Icons.now_widgets_outlined,color: Colors.grey.withAlpha(60),size: 30,)),
    ItemNavigationView(childAfter: const Icon(Icons.wifi,color: Colors.blue,size: 30,),childBefore: Icon(Icons.wifi_lock,color: Colors.grey.withAlpha(60),size: 30,)),
    ItemNavigationView(childAfter: const Icon(Icons.accessible,color: Colors.blue,size: 30,),
    childBefore: Icon(Icons.not_accessible,color: Colors.grey.withAlpha(60),size: 30,)),
    ],
),
```

In the above code, `onChangePage` is a callback function that is triggered when the user changes the
page. You can handle the page change logic inside this function.

The `curve` parameter allows you to specify the animation curve for the transitions between pages.
The `durationAnimation` parameter sets the duration of the animation.

The `items` parameter is a list of `ItemNavigationView` widgets. Each `ItemNavigationView`
represents a navigation item and consists of two child widgets: `childAfter` and `childBefore`.
These child widgets can be any widget, but in this example, we use `Icon` widgets from the `Icons`
class.

Feel free to customize the icons, colors, and other properties as per your needs.

## Example

For a complete example of using the NavigationView package, refer to
the [example](https://github.com/PuzzleTakX/navigation_view/tree/master/example) provided.

## License

This package is released under the MIT License. See the [LICENSE](https://github.com/PuzzleTakX/navigation_view/blob/master/LICENSE)
file for more details.

## Tags

    Flutter Navigation: Flutter, Mobile App Development, Navigation Component
        Emphasize the project's connection to Flutter and mobile app development.

    Customizable Navigation Bar: Custom Colors, Icons, Animated Transitions
        Highlight the project's customization options for navigation bars.

    User-Friendly Navigation: Smooth Transitions, User Experience, Animation Effects
        Indicate the improved user experience and animation features.

    Flutter Package: Flutter Plugin, Dart Library, UI Components
        Clarify that the project is a Flutter package or plugin.

    Mobile App UI: Mobile Interface, User Interface, App Design
        Attract users interested in enhancing their app's user interface.

    Open Source Project: GitHub Repository, Code Sharing, Collaboration
        Mention that the project is open source, encouraging collaboration.

    Customizable Icons: Icon Sets, Visual Design, Navigation Items
        Highlight the icon customization features.

    Flutter Animation: Flutter Animation Effects, Visual Enhancements
        Promote the use of Flutter's animation capabilities in the project.

    Flutter Development: Flutter Framework, App Development, UI Design
        Attract developers interested in Flutter development and UI design.

    App Navigation: App Navigation Bar, Page Transitions, UI Components
        Make it clear that the project is focused on app navigation and UI components.

## Credits

NavigationView is developed and maintained by [puzzleTak](https://github.com/PuzzleTakX).


CHANGELOG.md
# NavigationView (1.0.6)

[![pub package](https://img.shields.io/pub/v/navigation_view.svg)](https://pub.dev/packages/navigation_view)
[![pub points](https://img.shields.io/pub/points/navigation_view?color=2E8B57&label=pub%20points)](https://pub.dev/packages/navigation_view/score)



NavigationView is a Flutter package that provides a navigation component with customizable
colors and icons. It allows you to easily create a navigation bar with animated transitions between
different pages.
by most animation flutter

## Demo new

<img src="https://github.com/PuzzleTakX/navigation_view/blob/master/demo/3.gif?raw=true" alt="image_demo" width="300" height="700">

## Installation ☺

To use the NavigationView package, add the following dependency to your `pubspec.yaml`
file:

[![pub package](https://img.shields.io/pub/v/navigation_view.svg)](https://pub.dev/packages/navigation_view)
```yaml
dependencies:
navigation_view: ^1.0.6
```
Then, run `flutter pub get` to fetch the package.

## Usage

Import the package in your Dart file:

```dart
import 'package:navigation_view/navigation_view.dart';
```

Create a `NavigationView` widget and provide the necessary parameters:

```dart
 NavigationView(
onChangePage: (c) {},
curve: Curves.easeInBack,
durationAnimation: const Duration(milliseconds: 400),
items: [
ItemNavigationView(childAfter: const Icon(Icons.home_rounded,color: Colors.blue,size: 30,),
childBefore: Icon(Icons.home_outlined,color: Colors.grey.withAlpha(60),size: 30,)),
ItemNavigationView(childAfter: const Icon(Icons.widgets_rounded,color: Colors.blue,size: 30,),
childBefore: Icon(Icons.now_widgets_outlined,color: Colors.grey.withAlpha(60),size: 30,)),
ItemNavigationView(childAfter: const Icon(Icons.wifi,color: Colors.blue,size: 30,),childBefore: Icon(Icons.wifi_lock,color: Colors.grey.withAlpha(60),size: 30,)),
ItemNavigationView(childAfter: const Icon(Icons.accessible,color: Colors.blue,size: 30,),
childBefore: Icon(Icons.not_accessible,color: Colors.grey.withAlpha(60),size: 30,)),
],
),
```

In the above code, `onChangePage` is a callback function that is triggered when the user changes the
page. You can handle the page change logic inside this function.

The `curve` parameter allows you to specify the animation curve for the transitions between pages.
The `durationAnimation` parameter sets the duration of the animation.

The `items` parameter is a list of `ItemNavigationView` widgets. Each `ItemNavigationView`
represents a navigation item and consists of two child widgets: `childAfter` and `childBefore`.
These child widgets can be any widget, but in this example, we use `Icon` widgets from the `Icons`
class.

Feel free to customize the icons, colors, and other properties as per your needs.

## Example

For a complete example of using the NavigationView package, refer to
the [example](https://github.com/PuzzleTakX/navigation_view/tree/master/example) provided.

## License

This package is released under the MIT License. See the [LICENSE](https://github.com/PuzzleTakX/navigation_view/blob/master/LICENSE)
file for more details.

## Credits

NavigationView is developed and maintained by [puzzleTak](https://github.com/PuzzleTakX).

test/navigation_view_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:navigation_view/item_navigation_view.dart';
import 'package:navigation_view/navigation_view.dart';


void main() {
  testWidgets('NavigationView test', (WidgetTester tester) async {

    await tester.pumpWidget(SizedBox(
      width: double.maxFinite,height: 60,
      child: NavigationView(
        onChangePage: (c) {},
        curve: Curves.easeInBack,
        durationAnimation: const Duration(milliseconds: 400),
        items: [
          ItemNavigationView(
              childAfter: const Icon(
                Icons.home_rounded,
                color: Colors.blue,
                size: 30,
              ),
              childBefore: Icon(
                Icons.home_outlined,
                color: Colors.grey.withAlpha(60),
                size: 30,
              )),
          ItemNavigationView(
              childAfter: const Icon(
                Icons.widgets_rounded,
                color: Colors.blue,
                size: 30,
              ),
              childBefore: Icon(
                Icons.now_widgets_outlined,
                color: Colors.grey.withAlpha(60),
                size: 30,
              )),
          ItemNavigationView(
              childAfter: const Icon(
                Icons.wifi,
                color: Colors.blue,
                size: 30,
              ),
              childBefore: Icon(
                Icons.wifi_lock,
                color: Colors.grey.withAlpha(60),
                size: 30,
              )),
          ItemNavigationView(
              childAfter: const Icon(
                Icons.accessible,
                color: Colors.blue,
                size: 30,
              ),
              childBefore: Icon(
                Icons.not_accessible,
                color: Colors.grey.withAlpha(60),
                size: 30,
              )),
        ],
      ),
    ),);


  });
}


lib/navigation_view.dart
library navigation_view;


import 'package:flutter/material.dart';
import 'package:navigation_view/item_navigation_view.dart';

class NavigationView extends StatefulWidget{
  final Function(int) onChangePage;
  final Color? backgroundColor;
  final Color? borderTopColor;
  final Curve? curve;
  final Color? color;
  final BorderRadiusGeometry? borderRadius;
  final Gradient? gradient;
  final Duration? durationAnimation;
  final List<ItemNavigationView> items;
   const NavigationView({Key? key,


    required this.onChangePage,
    required this.items,
     this.durationAnimation,
    this.backgroundColor,
    this.borderRadius,
    this.gradient,
    this.color,
    this.curve,
    this.borderTopColor,


  }) : super(key: key);

  @override
  State<StatefulWidget> createState()=> _NavigationView();


}

class _NavigationView extends State<NavigationView>{

  final double borderRadius = 11.1;
  late int _currentPage = 0;


  final Color color = Colors.blue;
  Duration durationAnimation =  const Duration(milliseconds: 250);
  @override
  void initState() {
     if(widget.durationAnimation != null) durationAnimation = widget.durationAnimation!;
    _currentPage =  0;
    super.initState();
  }



  @override
  Widget build(BuildContext context) {
    return Container(width: double.maxFinite,color: (widget.backgroundColor != null) ? widget.backgroundColor : Colors.white,height: 55,
        child: Column(
          children: [
            Container(width: double.maxFinite,height: 1,
              color: (widget.borderTopColor != null) ? widget.borderTopColor! : Colors.grey.withAlpha(20)
              ,),
            Expanded(child: LayoutBuilder(
              builder: (p0, constraints){
                return Stack(
                  children: [
                    AnimatedPositioned(
                      curve: (widget.curve != null) ? widget.curve! : Curves.easeInOutQuint,
                      left: (constraints.maxWidth / widget.items.length) * ((widget.items.length - 1) - _currentPage),
                      width:  constraints.maxWidth / widget.items.length,
                      height: constraints.maxHeight,
                      duration: Duration(milliseconds: durationAnimation.inMilliseconds),
                      child: Padding(
                        padding: EdgeInsets.only(top: 5,bottom: 0,left: (45 /widget.items.length),right: (45 /widget.items.length)),
                        child:  InkWell(
                          onTap: (){

                          },

                          child: Container(
                            decoration: BoxDecoration(
                                color: Colors.transparent,
                                borderRadius: (widget.borderRadius != null) ? widget.borderRadius : BorderRadius.circular(borderRadius)
                            ),
                            child: Column(
                              children: [
                                Expanded(child: Padding(
                                  padding: const EdgeInsets.only(left: 5,right: 5),
                                  child: AnimatedContainer(
                                    duration: Duration(microseconds: durationAnimation.inMilliseconds ~/ 2.1),
                                    decoration: BoxDecoration(
                                      gradient:  (widget.gradient != null) ? widget.gradient! : LinearGradient(
                                          colors:  [
                                            Colors.white.withAlpha(0),
                                            (widget.color != null) ? widget.color!.withOpacity(0.2)  : color.withOpacity(0.2)
                                          ],


                                          begin: const FractionalOffset(0.0, 0.0),
                                          end: const FractionalOffset(0.0, 1.0),
                                          stops: const [0.0, 1.0],
                                          tileMode: TileMode.clamp),
                                    ),
                                  ),
                                )),
                                Padding(
                                    padding: const EdgeInsets.only(left: 4,right: 4),
                                    child: AnimatedContainer(
                                      duration: Duration(milliseconds: durationAnimation.inMilliseconds ~/ 2),
                                      width: double.maxFinite,height: 4,
                                      decoration: BoxDecoration(
                                          color: (widget.color != null) ? widget.color! :  color,
                                          borderRadius: (widget.borderRadius != null) ? widget.borderRadius! :  const BorderRadius.only(
                                            topLeft: Radius.circular(4),
                                            topRight: Radius.circular(4),
                                            bottomLeft: Radius.circular(0),
                                            bottomRight: Radius.circular(0),
                                          )
                                      ),
                                    ))
                              ],
                            ),

                          ),
                        ),
                      ),
                    ),
                    SizedBox(
                      width: double.maxFinite,
                      height: double.maxFinite,
                      child: Row(
                        textDirection: TextDirection.rtl,
                        children:widget.items.map((e) => Flexible(flex: 1,child: InkWell(
                          onTap: (){

                            setState(() {
                              _currentPage = widget.items.indexOf(e);
                              widget.onChangePage.call(widget.items.indexOf(e));
                            });
                          },
                          child: LayoutBuilder(
                              builder: (BuildContext context, BoxConstraints constraints) {
                                return  Center(

                                    child: AnimatedCrossFade(firstChild: e.childBefore, secondChild: e.childAfter, crossFadeState: (_currentPage == widget.items.indexOf(e)) ? CrossFadeState.showSecond : CrossFadeState.showFirst, duration:  Duration(milliseconds: durationAnimation.inMilliseconds ~/ 2))
                                );
                              }
                          ),
                        ),)).toList() ,
                      ),
                    ),

                  ],
                );
              },
            ),

            )
          ],
        )
    );
  }









}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:navigation_view/item_navigation_view.dart';
import 'package:navigation_view/navigation_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Navigation',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'PuzzleTak NavigationView'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Color> colors = [Colors.blue, Colors.red, Colors.amber, Colors.green];
  Color color = Colors.blue;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      backgroundColor: Colors.white.withAlpha(235),
      body: Padding(
        padding: const EdgeInsets.only(left: 10, right: 10),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            NavigationView(
              onChangePage: (c) {
                setState(() {
                  color = colors[c];
                });
              },
              durationAnimation: const Duration(milliseconds: 600),
              color: color,
              items: [
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.home_rounded,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.home_outlined,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.widgets_rounded,
                      color: Colors.red,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.now_widgets_outlined,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.wifi,
                      color: Colors.amber,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.wifi_lock,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.accessible,
                      color: Colors.green,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.not_accessible,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
              ],
            ),
            const SizedBox(
              height: 15,
            ),
            NavigationView(
              onChangePage: (c) {},
              curve: Curves.easeInBack,
              durationAnimation: const Duration(milliseconds: 400),
              items: [
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.home_rounded,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.home_outlined,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.widgets_rounded,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.now_widgets_outlined,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.wifi,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.wifi_lock,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.accessible,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.not_accessible,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
              ],
            ),
            const SizedBox(
              height: 15,
            ),
            NavigationView(
              onChangePage: (c) {},
              color: Colors.pink,
              curve: Curves.easeInQuint,
              durationAnimation: const Duration(milliseconds: 300),
              borderRadius: BorderRadius.circular(50),
              gradient: LinearGradient(
                  colors: [
                    Colors.white.withAlpha(0),
                    Colors.white.withOpacity(0.2)
                  ],
                  begin: const FractionalOffset(0.0, 0.0),
                  end: const FractionalOffset(0.0, 1.0),
                  stops: const [0.0, 1.0],
                  tileMode: TileMode.clamp),
              items: [
                ItemNavigationView(
                    childAfter: const Text(
                      "widget",
                      style: TextStyle(color: Colors.pink),
                    ),
                    childBefore: Container(
                      width: 60,
                      height: 20,
                      color: Colors.grey.withAlpha(50),
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.widgets_rounded,
                      color: Colors.pink,
                      size: 30,
                    ),
                    childBefore: Icon(
                      Icons.now_widgets_outlined,
                      color: Colors.grey.withAlpha(60),
                      size: 30,
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.wifi,
                      color: Colors.blue,
                      size: 30,
                    ),
                    childBefore: const Text(
                      "widget",
                      style: TextStyle(color: Colors.grey),
                    )),
              ],
            ),
            const SizedBox(
              height: 15,
            ),
            NavigationView(
              onChangePage: (c) {},
              color: Colors.deepPurpleAccent,
              curve: Curves.easeInCubic,
              durationAnimation: const Duration(milliseconds: 350),
              items: [
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.wifi,
                      color: Colors.deepPurpleAccent,
                      size: 25,
                    ),
                    childBefore: Text(
                      "Wifi",
                      style: TextStyle(color: Colors.grey.withAlpha(100)),
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.widgets_rounded,
                      color: Colors.deepPurpleAccent,
                      size: 25,
                    ),
                    childBefore: Text(
                      "Apps",
                      style: TextStyle(color: Colors.grey.withAlpha(100)),
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.alarm,
                      color: Colors.deepPurpleAccent,
                      size: 25,
                    ),
                    childBefore: Text(
                      "Alarm",
                      style: TextStyle(color: Colors.grey.withAlpha(100)),
                    )),
                ItemNavigationView(
                    childAfter: const Icon(
                      Icons.account_box,
                      color: Colors.deepPurpleAccent,
                      size: 25,
                    ),
                    childBefore: Text(
                      "Profile",
                      style: TextStyle(color: Colors.grey.withAlpha(100)),
                    )),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
